libs/cardano-ledger-binary/test-lib/Test/Cardano/Ledger/Binary/Twiddle.hs

```diff
@@ -201,17 +204,55 @@ instance Twiddle Term where
   twiddle v (TFloat x) = twiddle v x
   twiddle v (TDouble x) = twiddle v x
 
--- | Helper function for decoding an `Encoding` into a CBOR `Term`
 encodingToTerm :: Version -> Encoding -> Term
 encodingToTerm version enc =
   case decodeFull version (serialize version enc) of
     Right t -> t
     Left err -> error $ show err
 
--- | Helper function for converting an arbitrary value into a CBOR `Term`
 toTerm :: ToCBOR a => Version -> a -> Term
 toTerm version = encodingToTerm version . toCBOR
 
 -- | Wraps an arbitrary value into a `Twiddler`
 toTwiddler :: Twiddle a => Version -> a -> Gen (Twiddler a)
 toTwiddler v x = Twiddler v x <$> twiddle v x
+
+-- | Function for testing the invariant of a `Twiddle` instance. For a correct
+-- implementation, this property should always hold.
+twiddleInvariantProp :: forall a. Twiddle a => Version -> a -> Gen Property
+twiddleInvariantProp version x = do
+  t <- twiddle version x
+  let t' = encodingToTerm version $ toCBOR t
+  pure $ t === t'
+
+-- | Optional containers have two "empty" representations. One of
+-- them is to return an empty container and the other is to omit the field.
+-- This utility function randomly picks one of these representations, where
+-- omission is represented by `Nothing` and empty container is returned with
+-- `Just`. These values can then be easily concatenated with `catMaybes`.
+emptyOrNothing ::
+  forall t a b.
+  ( Foldable t,
+    Twiddle (t Void),
+    Monoid (t Void),
+    Twiddle (t b)
+  ) =>
+  Version ->
+  a ->
+  (a -> t b) ->
+  Gen (Maybe Term)
+emptyOrNothing v x f =
+  if null $ f x
+    then
+      oneof
+        [ Just <$> twiddle @(t Void) v mempty,
+          pure Nothing
+        ]
+    else Just <$> twiddle v (f x)
```

Comment:
This looks like overfunctionalization, instead this would be cleaner:

```suggestion
emptyOrNothing ::
  forall t b.
  (Foldable t, Twiddle (t Void), Monoid (t Void), Twiddle (t b)) =>
  Version ->
  t b ->
  Gen (Maybe Term)
emptyOrNothing v x =
  if null x
    then
      oneof
        [ Just <$> twiddle @(t Void) v mempty,
          pure Nothing
        ]
    else Just <$> twiddle v x
```


At call sites you just write instead of:
```haskell
cert' <- emptyOrNothing v txBody btbCerts
```

You write
```haskell
cert' <- emptyOrNothing v $ btbCerts txBody
```